### ISSUE TITLE
Provide a standard and fast way to execute a command inside an environment

### ISSUE THREAD

**Issue by gqmelo** [conda/conda#2379]
Currently if you want to run a command inside an environment you need to activate it and then execute the command.

This seems ok for interactive sessions, but if you want to integrate with external tools like IDE's and debuggers there is no standard way to do that.

A solution I came up with is to create a script that activates an environment and executes whatever is passed as argument. This works fine except that the activate scripts are very slow (most of time taking more than 0.5 seconds). I had to hack the activate to make it possible to use a script named `python` on my IDE.

Even IDE's like PyCharm, which has some conda support, does not actually activate the environment:

https://youtrack.jetbrains.com/issue/PY-18865

and if they want to do will probably have problems because of activate slowness.

Most of activate slowness is due to a lot of calls to conda.py
So what I propose is to have another script which only does the minimum necessary to run a command non-interactively:
- takes only a full path as the environment (this way you don't need to call conda.py to find where the envs dir is located)
- does not allow to be executed if another env is already activated (so it does not need to deactivate anything)
- modify the PATH
- execute the `activate.d` scripts

As it should probably not be source'd there is not need to exist a deactivate counterpart.

**Comment by msarahan** [comment]
Is this along the lines of what you're thinking: https://github.com/pelson/conda-execute ?

What you propose is also pretty much also what we do for Windows start menu shortcuts - except that there is no need to deactivate (it's a new shell), and we don't execute activate.d scripts.

**Comment by gqmelo** [comment]
It is actually not quite that, because conda-execute does more things, like declaring the dependencies and creating the environment.

I am talking more about given an existing environment, how can I run things inside it from an external tool like an IDE?

Nowadays it seems that people just open a shell, activate an env and open the IDE. I think this is far from ideal, because it makes difficult to deal with many env's (you would have to restart the IDE to switch).

Also, depending on what the activate.d scripts do you can have problems to execute the IDE on an activated environment. Let's say, for example, that one of these scripts change the PATH, so it changes the IDE behaviour. Or you have a java package, which breaks java based IDE's.
Unless the IDE or any other tool you want to use is also a conda package there is not guarantee that it will work on an activated env.

**Comment by msarahan** [comment]
Yes, I see.  This is a good idea.  There is http://conda.pydata.org/docs/commands/conda-run.html - but I have not heard of anyone using it, and I'm not sure what state it is in.

**Comment by tadeu** [comment]
Good idea indeed :).

`conda run -n env-name python` doesn't work, it says: `Error: App python not installed.`. Anyway, it tries to fetch package metadata before, so it would be slow.

Just brainstorm here, but it would be really nice if the official [Python Launcher](https://docs.python.org/3/using/windows.html#python-launcher-for-windows) supported running scripts inside conda environments without needing it to be activated (but anyway, the launcher is Windows only).

**Comment by tadeu** [comment]
Perhaps `conda run` could be revamped?

**Comment by gqmelo** [comment]
I did not know ```conda run```. But as tadeu said, it is slow for that case:

```bash
$ time conda run -n test echo
Fetching package metadata: ......
real	0m0.446s
user	0m0.352s
sys	0m0.032s
```

Note that IDE's like PyCharm makes dozens, maybe hundreds of calls to the interpreter, so the overhead should be really minimal. I don't think it is possible to do that with a conda command as every conda command at least open and parse the config file, which takes a lot of time.

Just to compare virtualenv activate is really fast:

```bash
$ time source ~/sentry_env/bin/activate

real	0m0.006s
user	0m0.004s
sys	0m0.000s
```

**Comment by gqmelo** [comment]
Just for the record, I already tried to optimize the activate script and the result was not good:

https://github.com/gqmelo/conda-env/commit/88345efa419b7771a1b9aae53b36f5713fffbf18

That is why my suggestion is to create a separate simple script to do just that

**Comment by msarahan** [comment]
Yeah, my hunch has been that the python scripts were the slow part, but it looks like you got rid of all of them (optionally ps1) in your experiment.  I'm not experienced in benchmarking shell scripts.  If anyone is, is airboats your comments here.

I'll be happy to review PRs for this, but I can't promise I'll be able to do this work myself any time soon.

**Comment by gqmelo** [comment]
>  I'm not experienced in benchmarking shell scripts. If anyone is, is airboats your comments here.

I am not either, but this gave me enough info:

```bash
export PS4='+ $(date "+%s.%N")   '
set -x
source activate test
```

> I'll be happy to review PRs for this, but I can't promise I'll be able to do this work myself any time soon.

Good to know, we (my company) are thinking of a quick solution now, but it should be probably similar to what we think could be officially supported by conda.
When we do that, we can discuss better over a working solution.

**Comment by gqmelo** [comment]
@msarahan in case you did not see yet, I created a solution for this:

https://anaconda.org/gqmelo/conda-wrappers

This basically uses another tool I created ([exec-wrappers](https://pypi.python.org/pypi/exec-wrappers)) to create wrappers for the environment it is being installed.

So if I install it on an environment called ```test``` I get a python wrapper on ```~/miniconda/envs/test/bin/wrappers/conda/python``` which I can use as if it were the real python interpreter.

I've found out an anaconda documentation explaining [how to use an IDE](https://docs.continuum.io/anaconda/ide_integration) with anaconda, but basically tells to point to the python interpreter, which does not cover cases when an activate is needed. So I think this is a very important feature to have a good integration with IDE's.

What do you think about something similar becoming a built-in feature? I was thinking about conda creating such wrappers whenever an environment is created/updated.

**Comment by msarahan** [comment]
@gqmelo, this looks like a very good thing, and you achieve very nice speedups.  Pinging @kalefranz about where this might fit in Conda's roadmap.

**Comment by kalefranz** [comment]
It seems like there might be two issues here. The first is that activate is slow. Definitely will dig into that and understand why. The second is to revamp conda run, or something similar. There are some other projects being worked on internally at Continuum (right now) that I'd like to see shake out and evolve first. Then we can figure out where conda run best fits.

**Comment by gqmelo** [comment]
> The first is that activate is slow. Definitely will dig into that and understand why.

I profiled once and it is because it makes a lot of calls to ```conda.py```, and ```conda.py``` needs to load the ```.condarc``` file. I don't think it is worth to optimize it, but even if you do, the other problem with ```activate``` is that it is intended to be used interactively.

And the same happens with ```conda-execute``` and ```conda-run```. They are nice tools, but you can use them only interactively. For example, if you want to get the python version with conda-run:

```bash
conda run -n test -- python --version
```

That is OK to run in an open terminal, but you cannot configure an IDE with this. And as IDE's normally can make hundreds of calls to commands like python, gcc, gdb, etc, every millisecond of overhead counts.

So my proposal is for conda to create wrappers for executables available inside the environments. And then any other external tool like IDE's, ```conda-run```, ```conda-execute``` could use these wrappers.

**Comment by jakirkham** [comment]
FYI `conda-run` is no more. ( https://github.com/conda/conda/issues/2682 ) Though it may come back.

**Comment by jakirkham** [comment]
What are the thoughts/recommendations on how we should solve this problem today?

**Comment by jakirkham** [comment]
Indeed. Am aware of that option. It feels a little overkill for what I'm trying to do though, but I guess it could work. Is it compatible with `conda` 4.4+ (am on 4.5 currently)? Also it there a way to limit which wrappers are generated? Only looking to wrap one executable instead of all of the ones in an environment.

**Comment by gqmelo** [comment]
> Is it compatible with conda 4.4+ (am on 4.5 currently)?

Yes, it should be. All it does is source the activation scripts and export some conda variables. Maybe some variable will not contain exact the same value (e.g. relative vs absolute path, etc.)

> Also is there a way to limit which wrappers are generated?

Yes, you can pass only the specific files you want separated by `:`:

```
create-wrappers  -t conda -d /tmp/conda_wrappers --conda-env-dir ~/miniconda3/envs/jupyter -f jupyter:python:qmake
```

So, just to add to the discussion, I've been using those wrappers for basically two use cases:

- Create wrappers for all executables in a particular environment (done by conda-wrappers package). Then point IDE's like `PyCharm` and `Eclipse+Pydev` to the `python` wrapper, Eclipse+CDT to gcc/gdb, etc.
- Install some tools (e.g ipyhon, jupyter) in separate environments and expose them as a wrapper in `~/bin`

**Comment by jakirkham** [comment]
To be clear, `conda` 4.4.0 changed how sourcing occurs. One use to run `source activate`/`activate`. Now one runs `conda activate`.

**Comment by jakirkham** [comment]
Interesting, thanks for the info, @gqmelo.

**Comment by kalefranz** [comment]
`source activate` / `activate` still works.  Hasn't been deprecated just yet.  We'll probably be adding in actual deprecation messages soon though, so that we can ultimately give `activate` back to virtualenv and not have clobber issues.

**Comment by jakirkham** [comment]
Good to know. Though have already migrated away from them in my own code at least. :)

Were you able to get tools like Jupyter to respect activation using this strategy, @gqmelo? I tried a primitive version of a shell wrapper script and Jupyter saw right through it, which caused it to skip activation in the notebook.

**Comment by gqmelo** [comment]
@jakirkham I don't use jupyter on a daily basis, just played with it a little bit. But I think it works fine. For example, after creating a jupyter wrapper in `/tmp/conda_wrappers/jupyter`, I'm able to run just `/tmp/conda_wrappers/jupyter notebook` on a new shell session and get an existing `*.ipynb` to open and run fine.

**Comment by mcg1969** [comment]
To be clear, this doesn't help if you need to it from the shell. Something like this works:
```
#!/bin/bash
_CONDA_EXE="$(cd "$(dirname "$0")"; pwd -P)"/conda
_ASK_CONDA=$($_CONDA_EXE shell.posix activate "$1") || \return $?
eval $_ASK_CONDA
shift
exec $@
```

**Comment by kalefranz** [comment]
For reference, a WIP `conda run` PR is at https://github.com/conda/conda/pull/7320

**Comment by ankostis** [comment]
For more reference, `conda run`(#7320) has been included in [*conda-4.6.0*]( https://github.com/conda/conda/blob/master/CHANGELOG.md)
to be released, release with sometime after June 2018.

Till then, and at my opinion, the next best thing is 
to install in your env @gqmelo's `conda-wrappers` recipe,
(implemented by #1205, as discussed in #2379), 
which clones all env-commands into self-contained executables (no activation needed) 
into a new sub-dir named `env/bin/wrappers`.

**Comment by agilebean** [comment]
I'm urgently waiting for `conda run`
but I saw conda 4.6 has not been released. 
Can you give a probable range for the released date?
Would be very very grateful...

**Comment by kalefranz** [comment]
Conda 4.6.0b0 is available now.

    conda config --add channels conda-canary
    conda update conda

**Comment by brettcannon** [comment]
@kalefranz I was checking out the PR and the help description seems to potentially conflicts with itself and the tests. If you look at https://github.com/conda/conda/blob/2de50fd699b62326bb706375c6d0c7b86da7a248/conda/cli/conda_argparse.py#L971 is says arguments to the command to be run should follow a `--`, but the example in the help and in the tests don't use the `--` separator. Is the use of `--version` meant to go to `conda` and not e.g. `python` or `sqlite3` as suggested in the help text or tests, respectively?

**Comment by joshburnett** [comment]
I was very happy to discover this functionality this week, as it looks perfect for my use case. However, I notice that when I execute a python script via 'conda run -n envname python scriptname.py', the various print statements don't get actually printed to the console until the entire script exits.

Any hints as to how I can make the output unbuffered? I already tried running my command with python unbuffered via the '-u' option, but no luck.

**Comment by kalefranz** [comment]
Another reason we should be using exec instead of subprocessing. See https://github.com/conda/conda/issues/8586

**Comment by mingwandroid** [comment]
No, I am afraid it is not.

Clearly conda run needs to be split into two, run and exec.

But that's got nothing to do with this problem. Here the problem is that we do not have an async thread to service conda run at all. Conda run is unfinished imho .. try doing something requiring input. That's an omission regardless of the execution method.

**Comment by NightMachinary** [comment]
That changes the environment after the command is done. It’s got zero
advantages over just switching the env and then running the command (in
fact it is just this.)

**Comment by infokiller** [comment]
One could run the command in a subshell to avoid changing the environment, for example in bash/zsh: `(conda activate py36 && python my_script.py)`. However, this still doesn't address the **speed** aspect (running `conda activate` is slow).

**Comment by connesy** [comment]
As far as I can tell, you can just use the python executable associated with the environment you want to run in:
```
$ conda activate dev
(dev) $ python -c "import numpy; print('Fine')"
Fine
(dev) $ ~/miniconda3/envs/test/bin/python -c "import numpy; print('Fine')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'numpy'
```

`conda run` could just point to the python path of the environment you want to run the command in. That would also eliminate the need to activate the environment first, significantly speeding up `conda run`.

**Comment by mingwandroid** [comment]
> That would also eliminate the need to activate the environment first, significantly speeding up conda run.

It would be fast, yes, but things would break.

**Comment by connesy** [comment]
> > That would also eliminate the need to activate the environment first, significantly speeding up conda run.
> 
> It would be fast, yes, but things would break.

Which things? I am currently using this solution in a Docker image, since I had problems with `conda run` not passing through logs from python.

**Comment by mingwandroid** [comment]
Well it will break any and all packages that require activation, clearly!

**Comment by mingwandroid** [comment]
Oops, didn't mean to close this.

**Comment by tadeu** [comment]
> As far as I can tell, you can just use the python executable associated with the environment you want to run in

For most packages, yes, but many packages require environment variables that are only defined in the activation, for example, GDAL: https://github.com/conda-forge/gdal-feedstock/blob/master/recipe/scripts/activate.sh

**Comment by karrtikr** [comment]
I can't use double dashes anymore
```
conda run -n test -- python --version
```
I have to run it like
```
conda run -n test python --version
```

Was the support for them removed?

**Comment by majidaldo** [comment]
> I can't use double dashes anymore
> 
> ```
> conda run -n test -- python --version
> ```
> 
> I have to run it like
> 
> ```
> conda run -n test python --version
> ```
> 
> Was the support for them removed?

same observation.
plus it seems like i have to be in the base environment.

**Comment by BrenBarn** [comment]
Curious about the status of this.  I think @gqmelo's remarks are spot-on.  People are increasingly using conda environments as a way to set up a "context" for executing individual tools: you have conda env for a particular tool, and you make you use that tool in that environment.  As gqmelo says, this begins to break down when you want use that environment-wrapped tool from within another tool.

Also, there are many meta-tools --- IDEs and things like Jupyter notebook --- that need to be able to run multiple Python instances simultaneously.  It's critical that these tools have a way to compartmentalize whatever needs to be compartmentalized so that each running Python instance behaves as if it were the only one.

[Here](https://github.com/ipython/ipykernel/issues/416) is an example of how the issue affects IPython.  [Here](https://github.com/conda/conda/issues/9587) is one about how it affects VSCode.  Some comments in those issues indicate related problems with other libraries (e.g., shapely/geos).

It seems like part of the problem is that conda is heavily reliant on the OS shell to set up its environment.  This means that tools that try to run conda without a shell (e.g., Python programs using `subprocess.Popen` with `shell=False`) fail.

It's clear to me from all these issues that there is a real need for a single-shot way to say "start up this conda environment and run this program with a single command which is not dependent on the shell".  In fact, I think we'd be a considerable way towards a solution if "this program" was only Python itself --- that is, just getting `conda run -n myenv python ...` to work so that you could run *Python* from a given environment without opening a shell.

It seems there is a good deal of mystery about what conda is actually doing to work its magic in the activation step.  For instance, [here](https://github.com/cylc/cylc-flow/issues/3704) is someone looking at the environment-variable changes wrought by `conda activate` and noting that it seems to be "too simple" to explain why the process takes so long.

Is there any documentation or working notes or anything like that that explains what all `conda activate` has to do to set up the environment, and how we could wrap that in to a single-shot command that doesn't require a separate activation step?  Besides setting environment variables (which can be done from all sorts of programs, not just shell scripts), what is it that conda needs to do in `activate`, and what are the challenges that prevent doing it in a single command?

**Comment by majidaldo** [comment]
> Curious about the status of this. I think @gqmelo's remarks are spot-on. People are increasingly using conda environments as a way to set up a "context" for executing individual tools: you have a conda env for a particular tool, and you make you use that tool in that environment. As gqmelo says, this begins to break down when you want use that environment-wrapped tool from within another tool.
> 
> Also, there are many meta-tools --- IDEs and things like Jupyter notebook --- that need to be able to run multiple Python instances simultaneously. It's critical that these tools have a way to compartmentalize whatever needs to be compartmentalized so that each running Python instance behaves as if it were the only one.
> 
> [Here](https://github.com/ipython/ipykernel/issues/416) is an example of how the issue affects IPython. [Here](https://github.com/conda/conda/issues/9587) is one about how it affects VSCode. Some comments in those issues indicate related problems with other libraries (e.g., shapely/geos).
> 
> It seems like part of the problem is that conda is heavily reliant on the OS shell to set up its environment. This means that tools that try to run conda without a shell (e.g., Python programs using `subprocess.Popen` with `shell=False`) fail.
> 
> It's clear to me from all these issues that there is a real need for a single-shot way to say "start up this conda environment and run this program with a single command which is not dependent on the shell". In fact, I think we'd be a considerable way towards a solution if "this program" was only Python itself --- that is, just getting `conda run -n myenv python ...` to work so that you could run _Python_ from a given environment without opening a shell.
> 
> It seems there is a good deal of mystery about what conda is actually doing to work its magic in the activation step. For instance, [here](https://github.com/cylc/cylc-flow/issues/3704) is someone looking at the environment-variable changes wrought by `conda activate` and noting that it seems to be "too simple" to explain why the process takes so long.
> 
> Is there any documentation or working notes or anything like that that explains what all `conda activate` has to do to set up the environment, and how we could wrap that in to a single-shot command that doesn't require a separate activation step? Besides setting environment variables (which can be done from all sorts of programs, not just shell scripts), what is it that conda needs to do in `activate`, and what are the challenges that prevent doing it in a single command?

*SUPER LIKE*!

This is why I've gotten VERY FAR in achieving this functionality with exec-wrappers thanks to @gqmelo . I actually solved [this](https://github.com/ipython/ipykernel/issues/416). I solve related issues using exec-wrappers and conda devenv.

On that note, 'my system' is something between conda build and conda devenv.

**Comment by oliver-sanders** [comment]
> If you would like this issue to remain open please:

Yes please, `conda activate` remains a major barrier to non-interactive uses.

> Is there any documentation or working notes or anything like that that explains what all conda activate has to do to set up the environment

It's worth taking a look at the activation script generated by conda-pack. It's quite minimal by comparison but appears to do the job.

This comment in the conda-build docs shows that a sufficiently large number of people are bypassing `conda activate` that it's worth warning feedstock contributors to avoid adding `activate.d` scripts to their projects:

> It is generally recommended to avoid using activate scripts when another option is possible because people do not always activate environments the expected way and these packages may then misbehave.
>
> -- https://docs.conda.io/projects/conda-build/en/latest/resources/activate-scripts.html#activate-scripts

**Comment by BrenBarn** [comment]
Agreed that, as far as I can tell, this isn't resolved.  It's just "stale" because it was never fixed.  It's still a problem, for all the same reasons listed in this issue thread.

**Comment by NightMachinery** [comment]
Any updates?

**Comment by imatiach-msft** [comment]
I was able to work around this by calling conda deactivate first, like this (and then passing to subprocess.check_call in my case):

        command = [
            "conda",
            "deactivate",
            "&&",
            "conda",
            "run",
            "-n",
            self.inference_conda_env,
            "python",
            get_wrapmodel_filepath(),
            "--model_path",
            self.model_path,
            "--wrapped_model_path",
            self.wrapped_model_path,
        ]

**Comment by lucazav** [comment]
I think we're all waiting for a solution to this problem. Any news from the Conda team? Or at least is there an intention to solve it?

**Comment by jezdez** [comment]
We've made progress on improving `conda run` indeed, in several changes over the past year. Is there anything you are looking for specifically that is blocking you @lucazav?

**Comment by NightMachinery** [comment]
Use `micromamba` instead. It provides this functionality, and unlike `conda`, it is not slow as hell. I have not noticed any missing conda API either.

```
micromamba run -n my_python_env mycommand
```

**Comment by sebma** [comment]
@NightMachinery Is micromamba a part of the [mamba package manager](https://github.com/mamba-org/mamba) ?

**Comment by oliver-sanders** [comment]
> @NightMachinery Is micromamba a part of the [mamba package manager](https://github.com/mamba-org/mamba) ?

Yes, mamba/micromamba ~= conda/miniconda.

> unlike conda, it is not slow as hell

I don't think mamba activate is much different to conda activate. I just ran a quick test and found that mamba activate/run is actually slightly slower in my environment (average of 5 calls):

```
/usr/bin/time <conda/mamba> activate <env>
/usr/bin/time <conda/mamba> run -n <env> python --version
```

|   | activate | run |
| --- | --- | --- |
| conda | 0.59s | 0.61s |
| mamba | 1.03s | 1.32s |

```
$ mamba --version 
mamba 1.5.1
conda 22.11.1
```

Both conda/mamba use Python to implement their CLI so it's not surprising their times are similar. I think mamba imports the conda CLI parser which would explain why it's marginally slower. By contrast, `conda pack` uses a shell script for activation, no Python, no imports, much faster.

**Comment by jezdez** [comment]
As an aside: mamba is a remarkable project that the conda maintainer team ([supported by Anaconda](https://www.anaconda.com/blog/a-faster-conda-for-a-growing-community)) has actively worked with in the past year to port into a conda plugin. Here's [the rollout plan to make the libmamba solver the default](https://conda.org/blog/2023-07-05-conda-libmamba-solver-rollout).

TLDR; conda will switch to using the conda-libmamba-solver as the default solver in the future. You can already use it by installing it in your base environment and configuring it NOW: https://conda.github.io/conda-libmamba-solver/getting-started/

```
conda install -n base conda-libmamba-solver
conda config --set solver libmamba
```

Anaconda Distribution, Miniconda and Miniforge already ship the package by default, you might already have it. The missing piece is to switch conda itself to use it by default (no config needed).

**Comment by NightMachinery** [comment]
@oliver-sanders micromamba is entirely in C IIRC. It’s a different client with less features (?).

**Comment by oliver-sanders** [comment]
> @oliver-sanders micromamba is entirely in C IIRC. It’s a different client with less features (?).

The solver (libmamba) is implemented in C++, but the command line interface is [implemented in python](https://github.com/mamba-org/mamba/blob/main/mamba/mamba/mamba.py).

It's a different client with **more** features (e.g. repo-query).

> has actively worked with in the past year to port into a conda plugin

Note, Conda is adopting the libmamba solver, however, mamba is more than just the solver.

**Comment by jezdez** [comment]
Thanks @oliver-sanders, regarding your point about conda-pack, its activation scripts are [here](https://github.com/conda/conda-pack/blob/main/conda_pack/scripts/posix/activate) (Posix) and [here](https://github.com/conda/conda-pack/blob/main/conda_pack/scripts/windows/activate.bat) (Windows). I'm not sure if you're suggesting doing anything with them, or if this was just informational.

Generally, it would be beneficial to understand where this ticket is going.

Could everyone still interested in this ticket verify if the current `conda run` suffices for their use cases? What specific issues have you found with it?

**Comment by oliver-sanders** [comment]
> I'm not sure if you're suggesting doing anything with them, or if this was just informational.

Just informational to demonstrate that the activation script can be implemented differently for performance gain. If you want a faster activation script, `conda pack` is currently your best option.

> it would be beneficial to understand where this ticket is going.

The OP is "Provide a **standard** and **fast** way to execute a command inside an environment".

* The **standard** part is now covered by `conda run` which is fine IMO.
* The **fast** bit is still outstanding.

Currently `<conda/mamba> actviate` adds >0.5s onto every command execution. On some filesystems, and when package activate scripts are involved, activation can take in excess of 1s. This presents a high penalty to deploying applications with Conda which is unfortunate.

The OP mentions text editor integrations which are a good example of an application where activation times are important. Other examples might include system utilities e.g. things like `coreutils` or regularly used command line applications e.g. developer tools.

Some systems have taken to bypassing `conda activate` and invoking scripts directly which works correctly for many applications. E.G. Python automatically looks in `$0/../lib/python<version>/site-packages` so calling an installed Python script by path without environment activation will usually work fine. However, this is a risky approach as the environment's `bin/` directory  is not in `PATH` (which could cause subprocess failures) and any package activation scripts are not run (which could cause a range of problems).

Despite the dangers of this approach, some applications are working this way to get around the Conda environment activation penalty (e.g. PyCharm as mentioned in the OP). The prevalence of this approach has resulted in the following warning in the conda-build docs:

> It is generally recommended to avoid using activate scripts when another option is possible because people do not always activate environments the expected way and these packages may then misbehave.
>
>    -- https://docs.conda.io/projects/conda-build/en/latest/resources/activate-scripts.html#activate-scripts

It would be good if we could come up with a faster approach to environment activation in Conda to avoid these issues.

**Comment by NightMachinery** [comment]
@oliver-sanders `mamba` is a different client from `micromamba`.

**Comment by oliver-sanders** [comment]
Gotcha, tried out micromamba, activate times are blazingly fast:

|   | activate | run |
| --- | --- | --- |
| conda | 0.59s | 0.61s |
| mamba | 1.03s | 1.32s |
| micromamba | 0.00s | 0.20s |

**Comment by lucazav** [comment]
> We've made progress on improving `conda run` indeed, in several changes over the past year. Is there anything you are looking for specifically that is blocking you @lucazav?

Thank you @jezdez, it was not clear from the thread that `conda run` was the suggested way to go.

**Comment by danyeaw** [comment]
Since conda run is now fully supported, and there are alternatives like micromamba run for those needing something higher performing, I'm going to go ahead and close this. Future work to improve how conda activates environments is being tracked by https://github.com/conda/conda/issues/14483.
